File: Server/src/services/resources/editor_context.py
from typing import Any

from fastmcp import Context
from pydantic import BaseModel

from models import MCPResponse
from models.unity_response import parse_resource_response
from services.registry import mcp_for_unity_resource
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry


class EditorContextSelection(BaseModel):
    """Summary of the current editor selection."""
    count: int = 0
    activeGameObject: str | None = None
    activeInstanceID: int = 0
    assetGUIDCount: int = 0


class EditorContextPrefabStage(BaseModel):
    """Whether a prefab stage is open and, if so, its asset path."""
    isOpen: bool = False
    assetPath: str | None = None


class EditorContextConsole(BaseModel):
    """Console counts by log type."""
    errors: int = 0
    warnings: int = 0
    logs: int = 0


class EditorContextData(BaseModel):
    """One-read situational-awareness snapshot fields."""
    editorState: Any | None = None
    selection: EditorContextSelection | None = None
    prefabStage: EditorContextPrefabStage | None = None
    console: EditorContextConsole | None = None


class EditorContextResponse(MCPResponse):
    """Union of editor state, selection, prefab stage, and console counts."""
    data: EditorContextData = EditorContextData()


@mcp_for_unity_resource(
    uri="mcpforunity://editor/context",
    name="editor_context",
    description="One-read situational-awareness snapshot: editor state, selection summary, prefab stage, and console counts by type. Read this once at task start instead of editor/state, editor/selection, and editor/prefab-stage separately.\n\nURI: mcpforunity://editor/context"
)
async def get_editor_context(ctx: Context) -> EditorContextResponse | MCPResponse:
    """Get a combined editor situational-awareness snapshot."""
    unity_instance = await get_unity_instance_from_context(ctx)
    response = await send_with_unity_instance(
        async_send_command_with_retry,
        unity_instance,
        "get_editor_context",
        {}
    )
    return parse_resource_response(response, EditorContextResponse)

File: MCPForUnity/Editor/Resources/Editor/EditorContext.cs
using System;
using System.Linq;
using System.Reflection;
using MCPForUnity.Editor.Helpers;
using MCPForUnity.Editor.Services;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEditor.SceneManagement;
using MCPForUnity.Runtime.Helpers;

namespace MCPForUnity.Editor.Resources.Editor
{
    /// <summary>
    /// One-read situational-awareness snapshot: the union of the editor state
    /// snapshot, the current selection summary, the open prefab stage, and the
    /// console counts by type. Lets an agent read a single resource at task start
    /// instead of four separate ones (get_editor_state + get_selection +
    /// get_prefab_stage + read_console).
    ///
    /// Wires up the <c>mcpforunity://editor/context</c> resource
    /// (Server/src/services/resources/editor_context.py) which dispatches the
    /// <c>get_editor_context</c> command name expected by this handler.
    ///
    /// Each sub-section is built defensively: a failure in one source degrades to
    /// a null sub-section rather than failing the whole read.
    /// </summary>
    [McpForUnityResource("get_editor_context")]
    public static class EditorContext
    {
        // Reflection into UnityEditor.LogEntries.GetCountsByType(ref int, ref int, ref int).
        private static readonly MethodInfo _getCountsByTypeMethod;

        static EditorContext()
        {
            try
            {
                Type logEntriesType = typeof(EditorApplication).Assembly.GetType("UnityEditor.LogEntries");
                _getCountsByTypeMethod = logEntriesType?.GetMethod(
                    "GetCountsByType",
                    BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
            }
            catch
            {
                _getCountsByTypeMethod = null;
            }
        }

        public static object HandleCommand(JObject @params)
        {
            try
            {
                var context = new JObject
                {
                    ["editorState"] = BuildEditorState(),
                    ["selection"] = BuildSelection(),
                    ["prefabStage"] = BuildPrefabStage(),
                    ["console"] = BuildConsole(),
                };

                return new SuccessResponse("Retrieved editor context.", context);
            }
            catch (Exception e)
            {
                return new ErrorResponse($"Error getting editor context: {e.Message}");
            }
        }

        private static JToken BuildEditorState()
        {
            try
            {
                return EditorStateCache.GetSnapshot();
            }
            catch
            {
                return JValue.CreateNull();
            }
        }

        private static JToken BuildSelection()
        {
            try
            {
                var active = UnityEditor.Selection.activeGameObject;
                var selectionInfo = new
                {
                    count = UnityEditor.Selection.count,
                    activeGameObject = active != null ? active.name : null,
                    activeInstanceID = UnityEditor.Selection.activeObject?.GetInstanceIDCompat() ?? 0,
                    assetGUIDCount = UnityEditor.Selection.assetGUIDs?.Length ?? 0,
                };
                return JObject.FromObject(selectionInfo);
            }
            catch
            {
                return JValue.CreateNull();
            }
        }

        private static JToken BuildPrefabStage()
        {
            try
            {
                var stage = PrefabStageUtility.GetCurrentPrefabStage();
                if (stage == null)
                {
                    return JObject.FromObject(new { isOpen = false, assetPath = (string)null });
                }

                return JObject.FromObject(new
                {
                    isOpen = true,
                    assetPath = stage.assetPath,
                });
            }
            catch
            {
                return JValue.CreateNull();
            }
        }

        private static JToken BuildConsole()
        {
            try
            {
                if (_getCountsByTypeMethod == null)
                {
                    return JValue.CreateNull();
                }

                var args = new object[] { 0, 0, 0 };
                _getCountsByTypeMethod.Invoke(null, args);

                return JObject.FromObject(new
                {
                    errors = (int)args[0],
                    warnings = (int)args[1],
                    logs = (int)args[2],
                });
            }
            catch
            {
                return JValue.CreateNull();
            }
        }
    }
}
